### PR TITLE
Add input plugin for s3

### PR DESCRIPTION
**Pull request by okkez**
This is work in progress for now.

How to use:
  (1) Create new SQS queue (use region same as S3)
  (2) Set proper permission to new queue
  (3) Set notification to S3 bucket
  (4) Run this plugin

### CODE REVIEW

**Comment by okkez** [comment]
TODO

* [ ] Add extractors same as compressors
* [ ] Write unit test
* [ ] Test in real world

**Comment by okkez** [comment]
This is my first input plugin, so I cannot write proper manner possibly.
I'm not sure about input plugin.
Request for comment to this PR, please.

**Comment by okkez** [comment]
Can we drop Ruby 1.9.3 support?

**Comment by okkez** [comment]
Drop Ruby 1.9.3 support in in_s3.rb.
I think new in_s3.rb does not need to support too old Ruby 1.9.3.

File: lib/fluent/plugin/in_s3.rb [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+      super
+
+      unless @sqs.queue_name
+        raise ConfigError, "sqs/queue_name is required"
+      end
+
+      begin
+        @extractor = EXTRACTOR_REGISTRY.lookup(@store_as).new(log: log)
+      rescue
+        $log.warn "#{@store_as} not found. Use 'text' instead"
```
Comment (repeatedly) [inline lib/fluent/plugin/in_s3.rb]: In s3 output, we use this fallback for backward compatibility.
I think raising error is better in s3 input.

File: lib/fluent/plugin/in_s3.rb [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+        # See http://docs.aws.amazon.com/sdkforruby/api/Aws/S3/Client.html
+      end
+      options
+    end
+
+    def create_s3_client
+      options = setup_credentials
+      options[:region] = @s3_region if @s3_region
+      options[:proxy_uri] = @proxy_uri if @proxy_uri
+      options[:s3_server_side_encryption] = @use_server_side_encryption.to_sym if @use_server_side_encryption
```
Comment (repeatedly) [inline lib/fluent/plugin/in_s3.rb]: Is this option needed in s3 input?

**Comment by repeatedly** [comment]
I think this plugin uses S3 event notification to sqs right?
Maybe, mentioning it in README is needed.

**Comment by okkez** [comment]
> I think this plugin uses S3 event notification to sqs right?
> Maybe, mentioning it in README is needed.

Yes, use s3 event.
I will update README about s3 event.

File: lib/fluent/plugin/in_s3.rb (hunk 1 above)
Comment (okkez) [inline lib/fluent/plugin/in_s3.rb]: OK. I will fix this.

**Comment by okkez** [comment]
Rebased.

**Comment by okkez** [comment]
Rebase again.

File: lib/fluent/plugin/in_s3.rb (hunk 2 above)
Comment (okkez) [inline lib/fluent/plugin/in_s3.rb]: Maybe, no.
When someone has read  access to encrypted object (by SSE) on S3 bucket, he can download decrypted object from S3 bucket.
Is that right?

**Comment by okkez** [comment]
Sorry :bow:
I closed this PR by mistake.

**Comment by okkez** [comment]
Finish working on this branch.

**Comment by tgalyean** [comment]
What is the status on this? Seems like nobody else is creating plugins which *pull* from S3. I think that functionality could be a major win for this project.

**Comment by repeatedly** [comment]
@okkez Could you rebase this patch? I will merge this patch and release rc version to test.

**Comment by okkez** [comment]
@repeatedly I've rebased and pushed.
Conflicts `.travis.yml` only.

**Comment by repeatedly** [comment]
thanks! merged

**Comment by ofirnn** [comment]
Hey - any news regards when it should be released? (0.7.0?)

**Comment by repeatedly** [comment]
@ofirnn Just released.

**Comment by repeatedly** [comment]
@okkez BTW, several input plugin tests failed with fleuntd v0.14. If you have a time, could you check this?

**Comment by okkez** [comment]
@repeatedly Sure. I will check soon.

**Pull request merged**